File: generate_mocks.py
import json
import re

sql_path = 'Inspirações/controle_ausencias_tabela_unica.sql'
with open(sql_path, 'r', encoding='utf-8') as f:
    sql = f.read()

admins = [
  {
    "id": "admin-1",
    "nome": "Lennon Administrador",
    "email": "admin@phdengenharia.eng.br",
    "senha": "968412",
    "perfil": "admin",
    "ativo": True
  }
]

gestores = []
usuarios = []
user_map = {}

lines = sql.split('\n')
inside_insert = False

for line in lines:
    if ') VALUES' in line:
        inside_insert = True
        continue
    t = line.strip()
    if inside_insert and t.startswith('('):
        # make it python evaluable tuple
        py_t = t.replace('DATE ', '')
        py_t = py_t.replace(' NULL,', ' None,')
        if py_t.endswith(','):
            py_t = py_t[:-1]
        try:
            row = eval(py_t)
            if len(row) >= 6:
                nome = str(row[0])
                email = row[1] if row[1] is not None else ''
                if email == '':
                    email = nome.replace(" ", "").lower() + "@phdengenharia.eng.br"
                
                funcao = str(row[2])
                lider_direto = str(row[3])
                superior = str(row[4])
                data_admissao = str(row[5])
                
                if nome not in user_map:
                    user_map[nome] = {
                        "nome": nome,
                        "email": email,
                        "senha": "123456",
                        "primeiroAcesso": True,
                        "funcao": funcao,
                        "lider_direto": lider_direto,
                        "superior": superior,
                        "dataAdmissao": data_admissao,
                        "ativo": True
                    }
        except Exception as e:
            pass

g_id = 1
u_id = 1

for nome, u in user_map.items():
    if u["nome"] == u["lider_direto"]:
        u["id"] = f"gestor-{g_id}"
        g_id += 1
        u["perfil"] = "gestor"
        gestores.append(u)
    else:
        u["id"] = f"user-{u_id}"
        u_id += 1
        u["perfil"] = "usuario"
        usuarios.append(u)

for u in usuarios:
    gestor = next((g for g in gestores if g["nome"] == u["lider_direto"]), None)
    if gestor:
        u["superiorId"] = gestor["id"]
    u["superior"] = u["lider_direto"]
    del u["lider_direto"]

for g in gestores:
    del g["lider_direto"]

mock_data_content = f"""// ===== DADOS MOCKADOS — Sistema de Gestão de Ausência =====

// Admins
export const admins = {json.dumps(admins, indent=2, ensure_ascii=False)};

// Gestores
export const gestores = {json.dumps(gestores, indent=2, ensure_ascii=False)};

// Usuários (colaboradores)
export const usuarios = {json.dumps(usuarios, indent=2, ensure_ascii=False)};

// Solicitações de Ausência
export const solicitacoes = [];

// Helper: buscar todos os usuários (admins + gestores + usuarios)
export const todosUsuarios = [...admins, ...gestores, ...usuarios];

export const autenticar = (email, senha) => {{
  return todosUsuarios.find(u => u.email === email && u.senha === senha && u.ativo);
}};

export const atualizarUsuario = (id, dados) => {{
}};

export const getColaboradoresPorGestor = (gestorId) => {{
  return usuarios.filter(u => u.superiorId === gestorId || u.id === gestorId);
}};

export const getSolicitacoesPorGestor = (gestorId) => {{
  return solicitacoes.filter(s => s.gestorId === gestorId);
}};

export const getSolicitacoesPorUsuario = (usuarioId) => {{
  return solicitacoes.filter(s => s.usuarioId === usuarioId);
}};

export const formatarData = (data) => {{
  if (!data) return '—';
  const d = new Date(data + 'T00:00:00');
  return d.toLocaleDateString('pt-BR');
}};

export const formatarMoeda = (valor) => {{
  if (!valor && valor !== 0) return '—';
  return valor.toLocaleString('pt-BR', {{ style: 'currency', currency: 'BRL' }});
}};

export const calcularPeriodoAquisitivo = (dataAdmissao) => {{
  const admissao = new Date(dataAdmissao + 'T00:00:00');
  const hoje = new Date();
  let inicioPA = new Date(admissao);

  while (true) {{
    const fimPA = new Date(inicioPA);
    fimPA.setFullYear(fimPA.getFullYear() + 1);
    fimPA.setDate(fimPA.getDate() - 1);
    if (fimPA >= hoje) {{
      return {{
        inicio: inicioPA,
        fim: fimPA,
        texto: `${{formatarData(inicioPA.toISOString().split('T')[0])}} - ${{formatarData(fimPA.toISOString().split('T')[0])}}`,
      }};
    }}
    inicioPA.setFullYear(inicioPA.getFullYear() + 1);
  }}
}};
"""

# Convert Python True/False to true/false in formatting
mock_data_content = mock_data_content.replace('True', 'true').replace('False', 'false')

with open('src/data/mockData.js', 'w', encoding='utf-8') as f:
    f.write(mock_data_content)

print(f"Mock data generated with {len(gestores)} gestores and {len(usuarios)} usuarios.")

File: generate_mocks.js
import fs from 'fs';

const sqlPath = 'Inspirações/controle_ausencias_tabela_unica.sql';
const sql = fs.readFileSync(sqlPath, 'utf8');

const admins = [
  {
    id: 'c2318237-b3f7-49ec-bb48-9d2a0c0c555d',
    nome: 'Admin',
    email: 'admin@phdengenharia.eng.br',
    senha: '968412',
    perfil: 'admin',
    ativo: true,
  }
];

const gestores = [];
const usuarios = [];
const ausencias = [];

const lines = sql.split('\n');
let insideInsert = false;
let userMap = {};

for (let line of lines) {
    if (line.includes(') VALUES')) {
        insideInsert = true;
        continue;
    }
    let t = line.trim();
    if (insideInsert && t.startsWith('(')) {
        // e.g. ('ADAILTON LUCAS GALINDO DE ANDRADE', 'adailton.andrade@phdengenharia.eng.br', 'CONSULTOR DE PLANEJAMENTO PL', 'TULIO RAFAEL MORAIS', 'LEONARDO AUGUSTO OLIVEIRA DRUMOND', DATE '2023-03-01', ...
        let content = t;
        if (content.endsWith('),') || content.endsWith(');')) {
            content = content.substring(1, content.length - 2);
        } else if (content.endsWith(')')) {
            content = content.substring(1, content.length - 1);
        }
        
        // Split by comma but ignore commas inside quotes
        let parts = content.split(/,\s*(?=(?:[^']*'[^']*')*[^']*$)/);
        
        if (parts.length >= 17) {
            const clean = (val) => {
                if (!val) return null;
                let v = val.trim();
                if (v === 'NULL' || v === 'None') return null;
                v = v.replace(/^DATE\s+/, '');
                v = v.replace(/^'|'$/g, '');
                return v;
            };

            let nome = clean(parts[0]);
            let email = clean(parts[1]);
            if (!email) email = `${nome.toLowerCase().replace(/\s+/g, '.')}@phdengenharia.eng.br`;
            let funcao = clean(parts[2]);
            let lider_direto = clean(parts[3]);
            let superior = clean(parts[4]);
            let data_admissao = clean(parts[5]);
            let ano_inicio = clean(parts[6]);
            let ano_fim = clean(parts[7]);
            let inicio_pa = clean(parts[8]);
            let fim_pa = clean(parts[9]);
            let limite_efetiva = clean(parts[10]);
            let saldo = clean(parts[11]);
            let qtd_programada = clean(parts[12]);
            let inicio_ausencia = clean(parts[13]);
            let fim_ausencia = clean(parts[14]);
            let dias_pendentes = clean(parts[15]);
            let status = clean(parts[16]);

            if (!userMap[nome]) {
                userMap[nome] = {
                    nome,
                    email,
                    senha: '123456',
                    primeiroAcesso: true,
                    funcao,
                    lider_direto,
                    superior,
                    dataAdmissao: data_admissao,
                    ativo: true
                };
            }

            ausencias.push({
                colaborador_nome: nome,
                ano_inicio,
                ano_fim,
                inicio_pa,
                fim_pa,
                limite_efetiva,
                saldo,
                qtd_programada,
                inicio_ausencia,
                fim_ausencia,
                dias_pendentes,
                status_original: status
            });
        }
    }
}

// Separate into gestores and usuarios
let gId = 1;
let uId = 1;
for (let nome in userMap) {
    let u = userMap[nome];
    // Check if user is a manager (if someone reports to them OR if they are their own leader)
    // Rule: Gestor is when they are their own leader direct OR if they appear as superior/lider_direto for others
    let isManager = (u.nome === u.lider_direto);
    
    if (isManager) {
        u.id = `gestor-${gId++}`;
        u.perfil = 'gestor';
        gestores.push(u);
    } else {
        u.id = `user-${uId++}`;
        u.perfil = 'usuario';
        usuarios.push(u);
    }
}

// Map superiorId and associate ausencias with user IDs
for (let u of [...gestores, ...usuarios]) {
    let gestor = gestores.find(g => g.nome === u.superior || g.nome === u.lider_direto);
    if (gestor) {
        u.superiorId = gestor.id;
    }
    
    // Link ausencias to this user
    ausencias.forEach(a => {
        if (a.colaborador_nome === u.nome) {
            a.colaborador_id = u.id;
        }
    });
}

const mockDataContent = `// ===== DADOS MOCKADOS — Sistema de Gestão de Ausência =====

// Admins
export const admins = ${JSON.stringify(admins, null, 2)};

// Gestores
export const gestores = ${JSON.stringify(gestores, null, 2)};

// Usuários (colaboradores)
export const usuarios = ${JSON.stringify(usuarios, null, 2)};

// Todas as Ausências (do SQL)
export const ausencias = ${JSON.stringify(ausencias, null, 2)};

// Solicitações de Ausência (Iniciadas no App)
export const solicitacoes = [];

// Helper: buscar todos os usuários (admins + gestores + usuarios)
export const todosUsuarios = [...admins, ...gestores, ...usuarios];

export const autenticar = (email, senha) => {
  return todosUsuarios.find(u => u.email === email && u.senha === senha && u.ativo);
};

export const getColaboradoresPorGestor = (gestorId) => {
  return usuarios.filter(u => u.superiorId === gestorId || u.id === gestorId);
};

export const getAusenciasPorColaborador = (colaboradorId) => {
  return ausencias.filter(a => a.colaborador_id === colaboradorId);
};

export const getStatusCalculado = (ausencia) => {
  const hoje = new Date();
  hoje.setHours(0, 0, 0, 0);
  
  const statusOriginal = ausencia.status_original;
  const limiteEfetiva = ausencia.limite_efetiva ? new Date(ausencia.limite_efetiva + 'T00:00:00') : null;
  const fimPA = ausencia.fim_pa ? new Date(ausencia.fim_pa + 'T00:00:00') : null;

  if (statusOriginal === 'Ausência Marcada') {
    return { label: 'Ausência Marcada', cor: 'var(--success)', status: 'aprovada' };
  }
  
  if (statusOriginal === 'OK') {
    return { label: 'OK', cor: 'var(--blue)', status: 'concluida' };
  }
  
  if (statusOriginal === 'Sem direito ainda') {
    if (fimPA) {
      const tresMesesAntes = new Date(fimPA);
      tresMesesAntes.setMonth(tresMesesAntes.getMonth() - 3);
      if (hoje >= tresMesesAntes && hoje <= fimPA) {
        return { label: 'Sem direito ainda (Aviso)', cor: 'var(--warning)', status: 'aviso', nota: 'Faltam menos de 3 meses para o fim do P.A.' };
      }
    }
    return { label: 'Sem direito ainda', cor: 'var(--gray-400)', status: 'bloqueada' };
  }
  
  if (statusOriginal === 'Marcação Pendente') {
    if (limiteEfetiva && limiteEfetiva < hoje) {
      return { label: 'Atrasado para marcar', cor: 'var(--error)', status: 'atrasada' };
    }
    return { label: 'Marcação Pendente', cor: 'var(--warning)', status: 'pendente' };
  }
  
  return { label: statusOriginal, cor: 'var(--gray-400)', status: 'desconhecido' };
};

export const formatarData = (data) => {
  if (!data) return '—';
  const d = new Date(data + 'T00:00:00');
  return d.toLocaleDateString('pt-BR');
};

export const formatarMoeda = (valor) => {
  if (!valor && valor !== 0) return '—';
  return Number(valor).toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' });
};
`;

fs.writeFileSync('src/data/mockData.js', mockDataContent, 'utf8');
console.log("Mock data generated with", gestores.length, "gestores,", usuarios.length, "usuarios and", ausencias.length, "ausencias.");
